src/endpoint/s3/ops/s3_post_object_select.js

```diff
@@ -0,0 +1,185 @@
+        md_params.part_number = part_number;
+    }
+    const object_md = await req.object_sdk.read_object_md(md_params);
+
+    s3_utils.set_response_object_md(res, object_md);
+    res.removeHeader('Content-Length'); // because transfer encoding is chunked
+    s3_utils.set_encryption_response_headers(req, res, object_md.encryption);
+    const obj_size = object_md.size;
+    const params = {
+        object_md,
+        obj_id: object_md.obj_id,
+        bucket: req.params.bucket,
+        key: req.params.key,
+        content_type: object_md.content_type,
+        noobaa_trigger_agent,
+        md_conditions,
+        encryption,
+    };
+    if (md_params.get_from_cache) {
+        params.get_from_cache = true;
+    }
+    if (part_number) {
+        params.part_number = part_number;
+    }
+    try {
+        const ranges = http_utils.normalize_http_ranges(
+            http_utils.parse_http_ranges(req.headers.range),
+            obj_size
+        );
+        if (ranges) {
+            // reply with HTTP 206 Partial Content
+            res.statusCode = 206;
+            params.start = ranges[0].start;
+            params.end = ranges[0].end;
+            const content_range = `bytes ${params.start}-${params.end - 1}/${obj_size}`;
+            dbg.log1('reading object range', req.path, content_range, ranges);
+            res.setHeader('Content-Range', content_range);
+            res.setHeader('Content-Length', params.end - params.start);
+            // res.header('Cache-Control', 'max-age=0' || 'no-cache');
+        } else {
+            dbg.log1('reading object', req.path, obj_size);
+        }
+    } catch (err) {
+        if (err.ranges_code === 400) {
+            // return http 400 Bad Request
+            dbg.log1('bad range request', req.headers.range, req.path, obj_size);
+            throw new S3Error(S3Error.InvalidArgument);
+        }
+        if (err.ranges_code === 416) {
+            // return http 416 Requested Range Not Satisfiable
+            dbg.warn('invalid range', req.headers.range, req.path, obj_size);
+            // let the client know of the relevant range
+            res.setHeader('Content-Range', 'bytes */' + obj_size);
+            throw new S3Error(S3Error.InvalidRange);
+        }
+        throw err;
+    }
+    // first_range_data are the first 4K data of the object
+    // if the object's size or the end of range is smaller than 4K return it, else get the whole object
+    if (params.object_md.first_range_data) {
+        const start = Number(params.start) || 0;
+        const end = params.end === undefined ? params.object_md.size : Math.min(params.end, params.object_md.size);
+        if (params.object_md.first_range_data.length >= end) {
+            const sliced_data = params.object_md.first_range_data.slice(start, end);
+            res.end(sliced_data);
+            return;
+        }
+    }
+
+    const read_stream = await req.object_sdk.read_object_stream(params, res);
+    if (read_stream) {
```

Comment:
what if not?

Reply:
I will end http response with status code 500 and nothing else.

Reply:
Something is still missing - did you try to run this through nsfs? 
If read_object_stream does not return a stream, we still need to pipe the s3select stream to the response.

Reply:
The concern I have with this current code is if combining the two separate pipeline calls works as expected - we never had this called this way - what we should verify is that propagating errors works well both from the reader to the response, and also in case of the client closing the connection then the response will raise errors and we should propagate it back to close our reader flows (either a stream or an async loop) and make sure we correctly handle aborting and discarding all the resources. It's hard to figure it all out by reading, so I would make a test for it.